microsoft/testsuites/cpu/common.py

```diff
@@ -0,0 +1,177 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+from __future__ import annotations
+
+from typing import Dict, List
+
+from lisa import BadEnvironmentStateException, Logger, Node
+from lisa.tools import Cat, Dmesg, Echo, Lscpu, Lsvmbus, Uname
+from lisa.util import SkippedException
+
+
+class CPUState:
+    OFFLINE: str = "0"
+    ONLINE: str = "1"
+
+
+def check_runnable(node: Node) -> None:
+    uname = node.tools[Uname]
+    kernel_version = uname.get_linux_information().kernel_version
+    config_path = f"/boot/config-{kernel_version}"
+    config_result = node.execute(f"grep CONFIG_HOTPLUG_CPU=y {config_path}", shell=True)
+    if config_result.exit_code != 0:
+        raise SkippedException(
+            f"the distro {node.os.name} doesn't support cpu hotplug."
+        )
+
+
+def set_vmbus_channels_target_cpu(
+    log: Logger, node: Node, target_cpu: str = "0"
+) -> Dict[str, str]:
+    uname = node.tools[Uname]
+    kernel_version = uname.get_linux_information().kernel_version
+    dmesg = node.tools[Dmesg]
+    lsvmbus = node.tools[Lsvmbus]
+    vmbus_version = dmesg.get_vmbus_version()
+    file_path_list: Dict[str, str] = {}
+    # the vmbus interrupt channel reassignment feature is available in 5.8+ kernel and
+    # vmbus version in 4.1+, the vmbus version is negotiated with the host.
+    if kernel_version >= "5.8.0" and vmbus_version >= "4.1.0":
+        # save the raw cpu number for each channel for restoring later.
+        channels = lsvmbus.get_device_channels(force_run=True)
+        for channel in channels:
+            for channel_vp_map in channel.channel_vp_map:
+                current_target_cpu = channel_vp_map.target_cpu
+                if current_target_cpu == target_cpu:
+                    continue
+                file_path_list[
+                    get_interrupts_assigned_cpu(
+                        channel.device_id, channel_vp_map.rel_id
+                    )
+                ] = current_target_cpu
+        # set all vmbus channel interrupts go into cpu target_cpu.
+        assign_interrupts(file_path_list, node, target_cpu)
+    else:
+        log.debug(
```

Comment:
If the channels cannot be changed, what's impact on the test? Should the test case be skipped?

Reply:
first try to set target cpu into 0
then get current idle cpu, if the first feature is not supported, we still get the idle cpu based on current VM real usage, then skip test or not, it is handled in method verify_cpu_hot_plug

Reply:
Add the explanation in code.

Reply:
I think it is already there

Reply:
> Add the explanation in code.
Below places are not enough?

https://github.com/microsoft/lisa/pull/1673/files#diff-cee6a3d2a6a62aabb5066517ea281c982068102c90d61f419a6f608d947cdc6aR37

https://github.com/microsoft/lisa/pull/1673/files#diff-d496653b77bd2eaf973ba7996aac876d2427107e3f880c4bb64b845c7aaae456R60

Reply:
1. The explanation should be align with the code. The test spec is good explanation, but it's better to mention it in the right code place too.
2. The second comments is good to explain it. But it should be after the `else`. It's weird at current place, because the next line is to return the list, it doesn't relate to the value is empty or not.

Reply:
updated